Automattic/wp-calypso
# Checkout: Merge full-credits payment method into free-purchase method

#### Proposed Changes

There are two ways to pay nothing during checkout: when the cart total is 0 (a "free purchase") and when the cart total is not 0 but the user has a number of credits equal to or greater than the cart total (a "full credits" purchase). Calypso checkout differentiates these by having two separate payment methods (`free-purchase` and `full-credits`), each with their own slightly different UI elements and payment processor functions. However, on the backend, there is only one payment method used for both, `WPCOM_Billing_WPCOM`.

In order to allow the backend more control over the payment methods being displayed (https://github.com/Automattic/payments-shilling/issues/1183) as well as to simplify the code needed to support these purchases in calypso, this PR merges these two payment methods into one. To do this, we alter the `free-purchase` payment method UX elements to display different text and alter the `free-purchase` processor function to delete tax location information for a full credits purchase.

#### Screenshots

Free purchase:

<img width="571" alt="Screen Shot 2022-10-25 at 6 41 29 PM" src="https://user-images.githubusercontent.com/2036909/197895915-2b0738a0-7144-4f73-9784-eb8713645870.png">

Full credits:

<img width="571" alt="Screen Shot 2022-10-25 at 6 43 01 PM" src="https://user-images.githubusercontent.com/2036909/197895930-37d38527-3f7a-414a-ab8e-0e2814ea732c.png">

#### Testing Instructions

There are automated tests covering both the UX and the payment processor, but you can also test manually.

- Test neither method
  - With no coupon and no credits, add a product to your cart and visit checkout.
  - Continue to the final step of checkout and verify that you see several payment methods, none of which mentions "Free" or "Credits".
- Test free payment method
  - Add a product to your cart and visit checkout.
  - Add a coupon to your cart that gives a 100% discount.
  - Continue to the final step of checkout and verify that you see the "Free Purchase" payment method.
  - Click to complete the purchase and verify that the purchase is successful.
- Test full credits payment method
  - Add enough credits to your account to completely cover the cost of a purchase.
  - Add a product to your cart and visit checkout.
  - Continue to the final step of checkout and verify that you see the "WordPress.com Credits" payment method.
  - Click to complete the purchase and verify that the purchase is successful.
